### ISSUE TITLE
Access to pages is granted regardless of role permissions

### ISSUE THREAD

**Issue by NeuralClone** [sulu/sulu#7265]
| Q | A
| --- | ---
| Sulu Version | 2.5.x
| Symfony Version | 6.3.6 or higher
| PHP Version | 8.1.27 and 8.2.15
| DB Version | MySQL 8.0.36
| Browser Version | N/A
| Related PR | [symfony#51986](https://github.com/symfony/symfony/pull/51986)

# Actual Behavior

If a security system is defined for a webspace, the frontend will serve protected pages to users even if their assigned roles don't have the required permissions.

# Expected Behavior

Access should be denied to users without the correct role permissions and a login prompt should be displayed to users that aren't yet authenticated.

# Steps to Reproduce

I've created a repository based on `sulu/skeleton` (v2.5.12) to reproduce this issue [here](https://github.com/NeuralClone/sulu-page-access-control-reproducer).

1. Install dependencies: `composer install`
2. Setup the database and load `/secure-area` page fixture: `php bin/adminconsole sulu:build dev`
3. Add a new role: `php bin/adminconsole sulu:security:role:create Member Website`
4. In the admin panel, activate the `Website` security system for `/secure-area` page and deny view access to all roles except for the newly created role.
5. Access `/secure-area` page without logging into the frontend.

Or you can manually reproduce it as follows:

1. Create and configure a new project using `sulu/skeleton` as outlined in the [Getting Started](https://docs.sulu.io/en/latest/book/getting-started.html) guide.
3. Add a security system to the `example` webspace and set `permission-check="true"`.
4. Enable user context caching as [described here](https://docs.sulu.io/en/latest/cookbook/user-context-caching.html).
5. In the admin panel, create a new user role for the webspace's security system.
6. Create a new page.
7. On the Permissions tab, enable the webspace's security system and deny access to all roles except the newly created role.
8. Publish the page.
9. Outside the admin panel, navigate to the newly created protected page to access it without logging in.

# Additional Information

This behavior occurs in both the `dev` and `prod` environments.

This only appears to impact pages that have permissions and not other content such as navigation or media. For example, if a page and an image have the same restricted permissions, the page is accessible without logging in but the image is not.

If a user doesn't have permission to view a page, that page is correctly hidden in its navigation contexts. But access is always granted if that user manually navigates to the path associated with that page.

This particular issue started appearing in Symfony 6.3.6 (probably in 5.4.29 as well since it was merged in 5.4) and seems to be directly related to [symfony#51986](https://github.com/symfony/symfony/pull/51986).

If I comment out the lines added in that PR, access to restricted pages isn't granted if the user doesn't have the right permissions. Otherwise, access is granted regardless of the user's permissions. That even applies to anonymous roles and pages requiring other roles the user doesn't have.

# Possible Solutions

I honestly have no idea as I don't know nearly enough about the inner workings of the components in question.

# Workarounds

Adding paths that should be secure to the `access_control` section of `security.yaml` takes priority over any security system defined for a webspace and prevents unauthorized users from accessing the pages in question. This kind of defeats the purpose of being able to secure specific pages through content creation, however.

For example:

```yaml
  access_control:
    # ...
    - { path: ^/secure-area, roles: ROLE_SULU_MEMBER }
``` 

Alternatively, it's possible to limit Symfony to version 6.3.5 or lower:

```json
    "extra": {
        "symfony": {
            "allow-contrib": true,
            "require": "6.3.*, <=6.3.5"
        }
    }
```

This isn't great due to the security vulnerabilities fixed in Symfony 6.3.8. But it does get permissions to work as expected.

**Comment by alexander-schranz** [comment]
I first really want to thank you for this detailed issue, I will have a look at it and keep you uptodate.

**Comment by NeuralClone** [comment]
You're welcome! Please let me know if there's anything I can do from my side to make things easier.

**Comment by alexander-schranz** [comment]
FYI will be part of the next release, which is planned for next week.

**Comment by NeuralClone** [comment]
That's fantastic. Thank you!

**Comment by alexander-schranz** [comment]
This is released now with 2.4.17 and 2.5.13 thx for the report.
